helios-client/src/main/java/com/spotify/helios/client/HeliosClient.java

```diff
@@ -344,20 +366,34 @@ private HttpURLConnection connect(final URI uri, final String method, final byte
     throw new TimeoutException("Timed out connecting to master");
   }
 
-  private HttpURLConnection connect0(final URI uri, final String method, final byte[] entity,
-                                     final Map<String, List<String>> headers)
+  private HttpURLConnection connect0(final URI ipUri, final String method, final byte[] entity,
+                                     final Map<String, List<String>> headers,
+                                     final String hostname)
       throws IOException {
     if (log.isTraceEnabled()) {
-      log.trace("req: {} {} {} {} {} {}", method, uri,
-                headers.size(),
+      log.trace("req: {} {} {} {} {} {}", method, ipUri, headers.size(),
                 Joiner.on(',').withKeyValueSeparator("=").join(headers),
                 entity.length, Json.asPrettyStringUnchecked(entity));
     } else {
-      log.debug("req: {} {} {} {}", method, uri, headers.size(), entity.length);
+      log.debug("req: {} {} {} {}", method, ipUri, headers.size(), entity.length);
     }
 
-    final URLConnection urlConnection = uri.toURL().openConnection();
+    final URLConnection urlConnection = ipUri.toURL().openConnection();
     final HttpURLConnection connection = (HttpURLConnection) urlConnection;
+
+    // We verify the TLS certificate against the original hostname since verifying against the
+    // IP address will fail
+    if (urlConnection instanceof HttpsURLConnection) {
+      ((HttpsURLConnection) connection).setHostnameVerifier(new HostnameVerifier() {
+        @Override
+        public boolean verify(String ip, SSLSession sslSession) {
+          final String tHostname = hostname.endsWith(".") ?
+                                   hostname.substring(0, hostname.length() - 1) : hostname;
+          return new DefaultHostnameVerifier().verify(tHostname, sslSession);
+        }
+      });
+    }
+
     connection.setRequestProperty("Accept-Encoding", "gzip");
```

Comment:
`Proxy.Type.HTTP` includes both HTTP and HTTPS proxies. The master endpoint will work fine as a "proxy" for HTTP, but not sure if it will forward HTTPS correctly. The nice thing about doing it this way would be that you could probably dump the hostname verification stuff.

But it might also not work at all. Up to you if you want to investigate or just punt on it.

I think long-term the best solution is to use the Apache HTTP client that lets you override DNS resolution and have <hostname> explicitly resolve to the single IP that we've selected.